compiler/rustc_builtin_macros/src/assert.rs

```diff
@@ -26,31 +26,41 @@ pub fn expand_assert<'cx>(
     // `core::panic` and `std::panic` are different macros, so we use call-site
     // context to pick up whichever is currently in scope.
     let sp = cx.with_call_site_ctxt(sp);
-    let tokens = custom_message.unwrap_or_else(|| {
-        TokenStream::from(TokenTree::token(
-            TokenKind::lit(
-                token::Str,
-                Symbol::intern(&format!(
-                    "assertion failed: {}",
-                    pprust::expr_to_string(&cond_expr).escape_debug()
-                )),
-                None,
-            ),
-            DUMMY_SP,
-        ))
-    });
-    let args = P(MacArgs::Delimited(DelimSpan::from_single(sp), MacDelimiter::Parenthesis, tokens));
-    let panic_call = MacCall {
-        path: Path::from_ident(Ident::new(sym::panic, sp)),
-        args,
-        prior_type_ascription: None,
+
+    let panic_call = {
+        if let Some(tokens) = custom_message {
+            // Pass the custom message to panic!().
+            cx.expr(
+                sp,
+                ExprKind::MacCall(MacCall {
+                    path: Path::from_ident(Ident::new(sym::panic, sp)),
+                    args: P(MacArgs::Delimited(
+                        DelimSpan::from_single(sp),
+                        MacDelimiter::Parenthesis,
+                        tokens,
+                    )),
+                    prior_type_ascription: None,
+                }),
+            )
+        } else {
+            // Pass our own message directly to $crate::panicking::panic(),
+            // because it might contain `{` and `}` that should always be
```

Comment:
It always contains `{}` (in `"assertion failed: {}"`), right?

Reply:
Nope. That's the format string it gives to `format!()` at compile time, inside rustc. It's the resulting string that is put in the macro expansion as string literal.

I could change that to do that at runtime, (produce the tokens `"assertion failed: {}", "stringified source here"`) but then const_panic will no longer work on `assert!(expr)`. See https://github.com/rust-lang/rust/pull/78069.

Reply:
My bad - I'm getting all of the various levels of `format!` mixed up :laughing: 

I also thought this was invoking the `panic!` macro - it's actually invoking the *function* `crate::panicking::panic`, which makes much more sense.